lib/dfu_target_uart_host/dfu_target_uart_host.c

```diff
@@ -0,0 +1,319 @@
+		ARRAY_SIZE(context.out_buffer),
+		&fnc,
+		&context.out_buf_cursor);
+	if (err) {
+		LOG_ERR("Could not encode response, error %d", err);
+		return;
+	}
+
+	if (context.out_buf_cursor == 0) {
+		LOG_DBG("Could not send error, no data in buffer");
+		return;
+	}
+
+	dfu_target_uart_host_send_out_buffer();
+}
+
+static void dfu_target_uart_host_send_offset(int error, size_t offset)
+{
+	int err;
+	int val = error == 0 ? (int) offset : error;
+
+	const struct DFU_UART_target_RemoteResp fnc = {
+		.type_choice = _DFU_UART_target_RemoteFnc_type_offset,
+		.error = val
+	};
+
+	err = cbor_encode_DFU_UART_target_RemoteResp(
+		context.out_buffer,
+		ARRAY_SIZE(context.out_buffer),
+		&fnc,
+		&context.out_buf_cursor);
+	if (err) {
+		LOG_ERR("Could not encode offset, error %d", err);
+		return;
+	}
+
+	if (context.out_buf_cursor == 0) {
+		LOG_DBG("Could not send error, no data in buffer");
+		return;
+	}
+
+	dfu_target_uart_host_send_out_buffer();
+}
+
+static void lte_uart_flush(const struct device *dev)
+{
+	uint8_t c;
+
+	while (uart_fifo_read(dev, &c, 1) > 0) {
+		continue;
+	}
+}
+
+static void uart_cb(const struct device *x, void *p)
+{
+	ARG_UNUSED(p);
+
+	int rx;
+	uint32_t ret;
+	static uint8_t read_buf[128];
+
+	while (uart_irq_update(context.uart_dev) &&
+		uart_irq_rx_ready(context.uart_dev)) {
+		rx = uart_fifo_read(x, read_buf, ARRAY_SIZE(read_buf));
+		if (rx > 0) {
+			ret = ring_buf_put(&context.rx_rb, read_buf, rx);
+			if (ret != rx) {
+				lte_uart_flush(context.uart_dev);
+				k_sem_give(&context.lte_uart_sem);
+				break;
+			}
+			k_sem_give(&context.lte_uart_sem);
+		}
+	}
+}
+
+static int
+dfu_target_uart_host_parse_init(const struct DFU_UART_target_RemoteFnc *fnc)
+{
+	context.file_size = fnc->data.file_size;
+	context.image_num = fnc->data.image_num;
+
+	return 0;
+}
+
+static int
+dfu_target_uart_host_parse_write(const struct DFU_UART_target_RemoteFnc *fnc)
+{
+	context.to_write_len = fnc->data.payload.len;
+	context.to_write = fnc->data.payload.value;
+
+	return 0;
+}
+
+static int
+dfu_target_uart_host_parse_done(const struct DFU_UART_target_RemoteFnc *fnc)
+{
+	context.download_success = fnc->data.success;
+
+	return 0;
+}
+
+static void execute_input_fnc(const struct DFU_UART_target_RemoteFnc *fnc)
+{
+	int err;
+	size_t offs_get;
+	static size_t progress;
+
+	LOG_DBG("DFU function called: %d", fnc->type_choice);
+
+	switch (fnc->type_choice) {
+	case _DFU_UART_target_RemoteFnc_type_init:
+		dfu_target_uart_host_parse_init(fnc);
+		progress = 0;
+		err = dfu_target_mcuboot_init(context.file_size,
+					      context.image_num,
+					      NULL);
+		dfu_target_uart_host_send_error(
+			_DFU_UART_target_RemoteFnc_type_init,
+			err);
+		break;
+	case _DFU_UART_target_RemoteFnc_type_write:
+		dfu_target_uart_host_parse_write(fnc);
+		err = dfu_target_mcuboot_write(context.to_write,
+					       context.to_write_len);
+		progress += context.to_write_len;
+		LOG_INF("Progress: %d %%", progress * 100 / context.file_size);
```

Comment:
If this function is called prior to the system being initialised, it's going to crash the device. There should be a proper state machine with state checking